File: simpleMult2.py
def multiply_matrices(first, second, result, r1, c1, r2, c2):
    for i in range(r1):
        for j in range(c2):
            # Initialize result[i][j] to 0 before summing
            result[i][j] = 0
            for k in range(c1):
                result[i][j] += first[i][k] * second[k][j]

def main(num_mult, mat_dim):
    first = [[i * j * 0.5 for j in range(mat_dim)] for i in range(mat_dim)]
    second = [[(i + j) * 0.5 for j in range(mat_dim)] for i in range(mat_dim)]
    result = [[None] for _ in range(mat_dim)] # All elements in result list replaced with None 

    for _ in range(num_mult):
        # Initialize result with None for each list in range
        for i in range(mat_dim):
            result[i] = [None]*mat_dim
            multiply_matrices(first, second, result, mat_dim, mat_dim, mat_dim, mat_dim)

    print("Result[5][5]:", result[5][5])

if __name__ == "__main__":
    import sys
    num_mult = int(100)
    mat_dim = int(1000)
    main(num_mult, mat_dim)

File: testing_programs/simpleMult2.c
#include <stdio.h>
#include <stdlib.h>

void multiplyMatrices(float* first, float* second, float* result, int r1, int c1, int r2, int c2)
{
  for (int i = 0; i < r1; ++i) {
    for (int k = 0; k < c1; ++k) {
      float temp = first[i * c1 + k]; // loop-invariant computation
      for (int j = 0; j < c2; ++j) {
        result[i * c2 + j] += temp * second[k * c2 + j];
      }
    }
  }
}

int main(int argc, char* argv[]) {

  int numMult = atoi(argv[1]);
  int matDim = atoi(argv[2]);
  int r1 = matDim;
  int c1 = matDim;
  int r2 = matDim;
  int c2 = matDim;

  float* first = (float*)malloc(r1 * c1 * sizeof(float));
  float* second = (float*)malloc(r2 * c2 * sizeof(float));
  float* result = (float*)calloc(r1 * c2, sizeof(float)); // initialize with zeros

  for (int i = 0; i < r1 * c1; i++) {
    first[i] = (i / c1) * (i % c1) * 0.5;
  }

  for (int j = 0; j < r2 * c2; j++) {
    second[j] = ((j / c2) + (j % c2)) * 0.5;
  }

  for (int i = 0; i < numMult; i++) {
    multiplyMatrices(first, second, result, r1, c1, r2, c2);
  }

  printf("Result[5][5]: %lf\n", result[5 * c2 + 5]);

  free(first);
  free(second);
  free(result);

  return 0;
}